x-pack/plugins/security_solution/public/common/hooks/use_form_with_warn/use_form_with_warn.ts

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+import { isEmpty } from 'lodash';
+import type { FormHook } from '../../../shared_imports';
+import { useForm, type FormConfig, type FormData } from '../../../shared_imports';
+import type { FormHookWithWarn } from './form_hook_with_warn';
+import { extractValidationResults } from './extract_validation_results';
+import type { ValidationResults } from './validation_results';
+
+export type FormWithWarnSubmitHandler<T extends FormData = FormData> = (
+  formData: T,
+  isValid: boolean,
+  validationResults: ValidationResults
+) => Promise<void>;
+
+interface FormWithWarnConfig<T extends FormData = FormData, I extends FormData = T>
+  extends Omit<FormConfig<T, I>, 'onSubmit'> {
+  onSubmit?: FormWithWarnSubmitHandler<T>;
+  options: FormConfig['options'] & {
+    warningValidationCodes: Readonly<string[]>;
+  };
+}
+
+interface UseFormWithWarnReturn<T extends FormData = FormData, I extends FormData = T> {
+  form: FormHookWithWarn<T, I>;
+}
+
+/**
+ * Form lib implements warning functionality via non blocking validators. `validations` allows to
+ * specify validation configuration with validator functions and extra parameters including
+ * `isBlocking`. Validators marked as `isBlocking` will produce non blocking validation errors
+ * a.k.a. warnings.
+ *
+ * The problem with the supported approach is lack of flexibility and necessary API like one for getting
```

Comment:
I shared the ticket. Will see how it could be tackled.

My goal was to move incrementally and the first step was existing functionality refactoring to reuse it in Prebuilt Rules Customization workflow. I'd be happy to reiterate on that when I get a chance.